feat: DSSE signing extension (#1371)
cloudevents/spec: cloudevents/extensions/verifiability.md

material MUST be transported and delivered along with the event that it
describes and not in separate events or even through different channels.

## Overview

This extension enables event producers to sign CloudEvents and consumers to
verify those signatures. The verification material is transported in an
extension attribute alongside the event data.

In a typical flow (using an SDK):
The producer passes a CloudEvent to the SDK, which creates the verification
material and adds it to the CloudEvent. When the consumer's CloudEvents SDK
receives a message with event and verification material, it performs a
verification of the signature against the key and passes on a verified event to
the consumer:

![An illustration showing how the producer's SDK signs an event, and the consumer's SDK verifies it](verifiability1.png)

The verification material is transported in an [Extension Context Attribute](https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#extension-context-attributes)
called `dssematerial` (see [Attributes](#attributes) section below).

The `dssematerial` binary value MUST be the UTF-8 encoding of a proper DSSE

> envelope:

Reviewer: DSSE JSON envelope (like below)

    JSON Envelope with:
    - `payloadType` of `"https://cloudevents.io/verifiability/dsse/v0.1"`
    - `payload` containing a Base64-encoded JSON object with:
      - `core`: Base64-encoded 32-byte SHA256 digest of the core CloudEvent fields (REQUIRED)
      - `ext`: Base64-encoded 32-byte SHA256 digest of the signed extension attribute values (OPTIONAL)
      - `signedextattrs`: JSON array of extension attribute name strings that were signed (OPTIONAL)
      - `ext` and `signedextattrs` MUST both be present or both be absent
      - The `signedextattrs` array MUST adhere to the following rules. If any
        of these rules are violated, the signing function MUST refuse to
        produce material and the verifier MUST discard the event:
        - Attribute names MUST NOT contain repetitions
        - Attribute names MUST NOT include any [core context attribute](https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#context-attributes) name (`id`, `source`, `specversion`, `type`, `datacontenttype`, `dataschema`, `subject`, `time`)
        - Attribute names MUST NOT include `dssematerial` (the verification material attribute)

>         - Extension attribute values MUST be of a supported CloudEvents type (`Boolean`, `Integer`, `String`, `Binary`, `URI`, `URI-reference`, or `Timestamp`)

Reviewer 2: I'm not sure what this means - there's no concept of an attribute which exists but isn't of a supported type.

If this is intended to mean that the signer *knows* the type of the attribute, that's slightly different. I think I'd probably talk about that separate from this bullet list though.

a JSON object created as follows:

```
CORE_DIGEST = SHA256(
  SHA256(UTF8(event.id)) +
  SHA256(UTF8(event.source)) +
  SHA256(UTF8(event.specversion)) +
  SHA256(UTF8(event.type)) +
  SHA256(UTF8(event.datacontenttype)) +
  SHA256(UTF8(event.dataschema)) +
  SHA256(UTF8(event.subject)) +
  SHA256(UTF8(RFC3339(UTC(event.time)))) +
  SHA256(event.data)
)

EXT_DIGEST = SHA256(
  SHA256(UTF8(CEString(event.extensionattr1))) +
  SHA256(UTF8(CEString(event.extensionattr2)))
)

VERIFICATION_MATERIAL = {
  "core": Base64(CORE_DIGEST),
  "ext":  Base64(EXT_DIGEST),          // omitted if no extension attributes are signed
  "signedextattrs": ["extensionattr1", "extensionattr2"]  // omitted if no extension attributes are signed
}
```

Notation used above:
- `+` denotes byte-sequence concatenation.
- `UTF8(s)` is the UTF-8 encoding of string `s`.
- `RFC3339(UTC(t))` is the RFC 3339 string representation of timestamp `t`
  normalized to UTC ("Zulu") time with second precision.
- `CEString(v)` is the CloudEvents canonical string encoding of an extension
  attribute value, as defined in step 5.2.1 of the [Signing Protocol](#signing-protocol)

>   (`Boolean`, `Integer`, `String`, `Binary`, `URI`, `URI-reference`, `Timestamp`).

Reviewer 2: Not sure what this list is intended for, but I think we can just remove the line (and move the period to the line above)

Author: It's mostly cause security typically wants things to be VERY clear. No way to interpret anything.

> - `event.data` is hashed as its raw byte representation — no UTF-8 wrapping

Reviewer 2: Hmm... this could be tricky. I've been focused on the representations of the attributes, but data is a whole other tricky story. What *is* the "raw byte representation" of an in-memory CloudEvent which has a JSON object as its data? If this is intended to be "the bytes that appear in the serialized form" then that gets us back to implementation difficulties.

(Historically, attributes have been *relatively* straightforward - we've had many, many discussions about the nature of data, particularly for event formats which can represents certain data types in a different way, e.g. where the data for a JSON-formatted event is itself a JSON object...)

Author: This only applies to event.data, not to context attributes. Context attributes use semantic canonicalization (Option 3 from our email thread), which you described as having “really significant benefits: Signature doesn’t change based on format, Much simpler implementation than either 1 or 2”, and that’s what the spec implements for all typed attributes (UTF8(s) for strings, RFC3339(UTC(t)) for timestamps, etc.).

For event.data, raw bytes are unavoidable: there is no CE type system for event data, so we cannot apply semantic canonicalization. The note at lines 327–328 calls this out explicitly, any intermediary that reserializes event data (whitespace changes, key reordering) must re-sign. This is the same trade-off you acknowledged in the email when discussing Option 2: “That’s ‘only’ inefficient rather than particularly difficult, but it does mean that support for this extension needs to be built into every event formatter.” We avoided that for attributes via Option 3; for data there is no equivalent.

  is applied, because event data is not necessarily text (see Case 5 for a
  binary example).
- If an OPTIONAL Context Attribute or a signed extension attribute is not
  set on the event, the SHA-256 of the empty byte sequence is used in its
  place. See the Security Considerations section for the limitation this
  introduces.

`CORE_DIGEST` is the digest of the concatenated digest list of the mandatory
Context Attributes, the OPTIONAL Context Attributes, and the event data.
`EXT_DIGEST` is the digest of the concatenated value digest list of the
signed extension attributes. The extension attribute names are authenticated
by being included in the `signedextattrs` array inside the signed container.
The `ext` and `signedextattrs` fields are only present when extension
attributes are signed.

#### Signing Protocol

This is how to sign a CloudEvent using DSSE:

1. choose a signing key

> 2. choose the list of "signed extension attributes" (the list MUST adhere to the constraints defined in the [`dssematerial`](#dssematerial) attribute definition: no repetitions, no Context Attribute names, and no `dssematerial`)

Reviewer 2: ```suggestion
2. choose the list of "signed extension attributes" (the list MUST adhere to the constraints defined in the [`dssematerial`](#dssematerial) attribute definition: no repetitions, no Core Context Attribute names, and no `dssematerial`)
```

    4. compute the SHA256 digest of the event's [`type`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#type) Context Attribute in UTF8 and append it to the byte sequence *(if the attribute is not set, use the digest of the empty byte sequence)*
    5. compute the SHA256 digest of the event's [`datacontenttype`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#datacontenttype) Context Attribute in UTF8 and append it to the byte sequence *(if the attribute is not set, use the digest of the empty byte sequence)*
    6. compute the SHA256 digest of the event's [`dataschema`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#dataschema) Context Attribute in UTF8 and append it to the byte sequence *(if the attribute is not set, use the digest of the empty byte sequence)*
    7. compute the SHA256 digest of the event's [`subject`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#subject) Context Attribute in UTF8 and append it to the byte sequence *(if the attribute is not set, use the digest of the empty byte sequence)*

>     8. compute the SHA256 digest of the event's [`time`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#time) Context Attribute normalized to RFC 3339 Zulu format with second precision (no subsecond component) and append it to the byte sequence (if the attribute is not set, use the digest of the empty byte sequence)

Reviewer: "Zulu" is mentioned once in RFC3339, but it will be helpful to show a formatted datetime example of what you mean here and also link to the RFC. That's a good potential source of hours of debugging fun if one misses this detail.

       Note: Time normalization to Zulu format with second precision ensures signature verification remains valid even when intermediaries deserialize and reserialize events with different timezone representations or subsecond precisions of the same timestamp.
    9. compute the SHA256 digest of the event's [`data`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#event-data) and append it to the byte sequence
4. compute `CORE_DIGEST` as the SHA256 digest of the byte sequence from step 3
5. if the list from step 2 is not empty:
    1. create an empty byte sequence for the ext digest
    2. for each extension attribute in the list from step 2 (in the given order, from lowest to highest index)
        1. compute the SHA256 digest of the extension attribute's value and append it to the byte sequence. Extension attribute values MUST be serialized using their CloudEvents canonical string encoding before hashing:
            - `Boolean`: "true" or "false" (case-sensitive)
            - `Integer`: decimal representation without leading zeros, fraction, or exponent (per RFC 7159 Section 6)

>             - `String`: the string value as-is

Reviewer 2: A string value can't be hashed, as hashing is an operation on binary data. I suspect we want to say you hash the UTF-8-encoded form of the string.

            - `Binary`: Base64 encoding per RFC 4648
            - `URI`: the absolute URI string per RFC 3986 Section 4.3
            - `URI-reference`: the URI-reference string per RFC 3986 Section 4.1

>             - `Timestamp`: RFC 3339 Zulu format with second precision (normalized to UTC as described in step 8)

Reviewer 2: Again, I'd separate out "here's how you hash a value" (on a per type basis) from the uses of it. That way you don't need any repetition, and key implementation details (e.g. how to hash a timestamp value) aren't buried in the list.

Attributes that were signed by the producer. Any extension attributes not listed
in `signedextattrs` MUST NOT appear on the verified event returned to the
consumer. If a consumer needs access to unsigned extension attributes, it MUST
explicitly obtain them from the original pre-verification event and MUST clearly
distinguish them as unverified data not covered by the signature. This ensures
that consumers cannot accidentally treat unverified extension attribute values as
verified without deep inspection of which attributes were signed.

### What's verifiable and what isn't?

Depending on how a CloudEvent is transported in a [binary-mode, structured-mode or batch-mode](https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#message)
CloudEvent message different data can be verified:

|Verifiable information	|binary-mode	|structured-mode	|batch-mode	|comment	|
|---	|---	|---	|---	|---	|
|Event Data (payload)	|✅	|✅	|✅	|	|
|REQUIRED Context Attributes	|✅	|✅	|✅	|	|
|OPTIONAL Context Attributes	|✅	|✅	|✅	|See notes below for time attribute	|
|Extension Context Attributes	|✅	|✅	|✅	|Optional (per attribute)	|
|Metadata added by transports	|❌	|❌	|❌	|	|

*Notes:*

> * *In [CloudEvent's type system](https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#type-system) a `Timestamp`'s string encoding is [RFC 3339](https://tools.ietf.org/html/rfc3339). This means that verification of the `time` Context Attribute can only be done with second precision, even though an SDK might allow passing in a timestamp with nanosecond precision.*

Reviewer 2: It's not RFC3339 which limits us to second precision - it's interoperability. We don't know what degree of precision any given consumer will use, so it's simplest to truncate to the second.

Author: Good call out. I remember there's a timestamp RFC that is specifically second precisions, but I may be misremembering or blurring RFCs.

But good callout

> * *In [CloudEvent's official Protocol Buffers format](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/formats/cloudevents.proto#L57), the `time` Context Attribute is encoded as a `google.protobuf.Timestamp` and hence does not include time zone information (which RFC 3339 would allow). For interoperability with CloudEvent setups using the Protocol Buffers format, time zone information is ignored in the signing and verification process.*

Reviewer: We should assume that all timezone-less datetime values are UTC.

> * *The signature covers the exact byte representation of the event data. Intermediaries that deserialize and reserialize event data (e.g., reformatting JSON whitespace or reordering keys) will invalidate the signature. Deployments requiring end-to-end verifiability SHOULD avoid such intermediaries or accept re-signing at trust boundaries.*

Reviewer 2: This comes back to what "the exact byte representation of the event data" is, and how tricky that could be for implementations...

Author: Same answer as above

This applies only to event.data, not to context attributes. For context attributes the spec uses semantic canonicalization (Option 3 from our March 11 discussion), which you preferred precisely because it avoids the wire-bytes implementation pain you described: “often the bytes that were on the wire aren’t known by the time you’re dealing with the parsed representation... the JSON event formatter reads the whole body as a document. At that point, the original bytes on the wire have gone.”

For event.data we cannot apply the same trick, there is no CE type for data. Signing requires stable bytes, and the spec documents the consequence clearly (lines 327–328): intermediaries that reserialize data must resign. SDK authors should sign as close to the wire as possible. Happy to add a note in the spec making this explicit for SDK implementors.

Reviewer 2: Right - sorry for missing this before. I think in many cases this may make implementation close to infeasible, unfortunately.

For example in https://github.com/cloudevents/sdk-csharp/blob/main/src/CloudNative.CloudEvents.SystemTextJson/JsonEventFormatter.cs#L138 (and similar event formatters) the first thing that's done is to read the event as a JSON document. At that point, we simply don't have the original bytes. So to read the wire data, we'd probably have to maintain our own *very* low-level JSON parser in order to get the raw bytes.

I hate to be a downer, but I think for at least a lot of languages, this will simply break any layered architecture of the SDK. It may be feasible in some other languages of course.

Author: For event.data I want to acknowledge you raised this in our March email too, and at the time you said "That feels somewhat reasonable as a spec, but a real pain to implement in an SDK." We proceeded on that basis, and the spec documents the consequence clearly. So I dont think it was missed, per se.

On the C# SDK example: the formatter reading the whole JSON document before the raw bytes are available is a real constraint, but the solution is to capture the bytes during parsing rather than after. System.Text.Json exposes raw UTF-8 bytes for a value node via GetRawText() or an equivalent reader-level API, no custom low-level parser needed. The SDK would need to thread this through the signing path, which is additional work, but it's the same class of work any checksum/integrity system requires.

The alternative, defining a canonical serialization for event.data, would require specifying rules for every possible content type (JSON, XML, binary, protobuf, etc.), which is far out of scope. Raw bytes is the only content-type-agnostic option.

Producers typically sign before serialization where raw bytes are always available. For SDKs that receive-and-propagate, capturing bytes at the formatter layer is the right integration point. We can capture that in the spec if needed

Reviewer 2: > but the solution is to capture the bytes during parsing rather than after. System.Text.Json exposes raw UTF-8 bytes for a value node via GetRawText() or an equivalent reader-level API, no custom low-level parser needed.

I may have chosen a bad example - it's *possible* that we can do this for System.Text.Json without very many changes. I think I'd need to check it really carefully for every format - for example, in protobuf it may be feasible due to parsing effectively preserving everything we need.

At the very least, I think it'll mean work on a "per language, per format" basis, and in a way which makes the implementation significantly more complex. That may well be unavoidable, and I acknowledge that there's probably no better approach - but with our current state of SDKs being "it's hard to find maintainers at all" that may well mean we end up with an extension that is *very* spottily implemented :(

Author: As someone who used to maintain SDKs, I understand the pain :(.

I will have to check with legal, but we may be able to help with some implementations, e.g., golang, java, python.

C# is a little out of skillset though ><. At this point it seems like figuring out the best path to get this implemented.

I will have a commit that updates the other points, e.g., the duplicated canonicalization stuff, type lists, etc, soon

Through this process consumers can be sure that events were in fact produced by the claimed producer (authenticity), and that the events were received exactly as they were sent, and not modified in transit (integrity)—without needing to trust any intermediaries.

The threats addressed by this proposal are those of malicious actors impersonating CloudEvent producers and of malicious actors modifying messages in transit.

With interoperability in mind, this design opts for simplicity and robustness wherever possible.

## Non-goals

This proposal only applies to individual events.
It does not give consumers any guarantees about the completeness of the event stream or the order that events are delivered in.

The threats of a malicious actor preventing events from being delivered or swapping their order are not addressed by this proposal.
Neither are the possibilities of events being accidentally dropped, delivered in the wrong order or the same event being delivered multiple times.
These challenges can be addressed through means of adding the necessary information inside the event payloads.

> Because the CloudEvents specification [requires](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#id) the combination of event `source` and `id` to be unique per event, signature replays for identical events are not considered.

Reviewer 3: is "signature replays" a "security" phrase? I wonder if everyone will know what's being said here. I'm not sure I do :-)     Is it saying "we don't guarantee that replays will look the same", or is it saying "different events with the same id/souce might look the same even if other data is different" ? Or something else?

Further, this proposal only aims at *verifiability*.
It does not aim to enable *confidentiality*.
Consequently, it does not address the threat of unauthorized parties being able to read CloudEvents that were not meant for them (see [Privacy & Security](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#privacy-and-security) in the CloudEvents spec).

While the design in this proposal *can* be used by authorized intermediaries to modify and re-sign events, it explicitly does not aim to provide a cryptographic audit trail of event modifications.

As a general principle, this proposal aims to avoid cryptographic agility in favor of simplicity.

## Constraints

We have set the following constraints for the proposed design:

> **Verifiability MUST be OPTIONAL:** This ensures that the additional burden of producing verification material and performing verification only applies when verifiability is desired, which is not always the case.

Reviewer 3: I wonder if it's clear that just because a producer adds the extra verifiable stuff to the CE, the consumer is not required to do anything with it or even know what those extensions mean? Could just be me.

Consumers that do not support verification can consumer signed events as if they were unsigned.

**The verification material MUST be contained in the same message as the event:** The design aims to be simple and robust, and so the verification material MUST be transported and delivered along with the event that it describes—and not in separate events or even through different channels.

## Overview

The producer passes a CloudEvent to the SDK, which creates the verification material and adds it to the CloudEvent. When the consumer’s CloudEvents SDK receives a message with event and verification material, it performs a verification of the signature against the key and passes on a verified event to the consumer:

![An illustration showing how the producer's SDK signs an event, and the consumer's SDK verifies it](verifiability1.png)

The verification material is transported in an [Extension Context Attribute](https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#extension-context-attributes) called `dssematerial`:

> * Type: `String`

Reviewer 3: You may want to follow the pattern in other extensions of having an "Attributes" section - see: https://github.com/cloudevents/spec/blob/52387e31cc41688ba1ce56ec8f040554d3517592/cloudevents/extensions/authcontext.md#attributes

* Description: The [DSSE JSON Envelope](https://github.com/secure-systems-lab/dsse/blob/master/envelope.md) that can be used to verify the authenticity and integrity of CloudEvent.
* Constraints:
    * OPTIONAL
    * If present, MUST be Base64 encoded

> The verification material, once Base64 decoded, looks something like this:

Reviewer 3: "looks something like" isn't precise enough. Are all of those 5 fields required? If so, then say "MUST adhere to the following form". If the fields can change based on the mechanism used to do the verifying then we'll need to be more creative, so let's discuss because across all mechanisms/formats there MUST be at least one consistent field (payloadType I assume) that people can rely upon to disambiguating them.

    4. compute the SHA256 digest of the event's [`type`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#type) Context Attribute in UTF8 and append it to the byte sequence
    5. compute the SHA256 digest of the event's [`datacontenttype`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#datacontenttype) Context Attribute in UTF8 and append it to the byte sequence *(if the attribute is not set, use the digest of the empty byte sequence)*
    6. compute the SHA256 digest of the event's [`dataschema`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#dataschema) Context Attribute in UTF8 and append it to the byte sequence *(if the attribute is not set, use the digest of the empty byte sequence)*
    7. compute the SHA256 digest of the event's [`subject`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#subject) Context Attribute in UTF8 and append it to the byte sequence *(if the attribute is not set, use the digest of the empty byte sequence)*

>     8. compute the SHA256 digest of the event's [`time`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#time) in RFC 3339 Zulu format and append it to the byte sequence *(if the attribute is not set, use the digest of the empty byte sequence)*

Reviewer 3: Meaning, add the note text to this bullet with a MUST  ??

4. parse the result into a [JSON DSSE Envelope](https://github.com/secure-systems-lab/dsse/blob/master/envelope.md)
    1. if parsing fails, the verification material is corrupted and the event MUST be discarded
5. follow the [DSSE verification protocol](https://github.com/secure-systems-lab/dsse/blob/master/protocol.md#protocol)
    1. filter signatures by `keyid` from  list of acceptable signing keys
    2. if verification fails, the event MUST be discarded
6. read the envelope’s `payloadType` value
    1. if it is not equal to `https://cloudevents.io/verifiability/dsse/v0.1`, the verification payload is unknown and the event MUST be discarded
7. read the envelope’s `payload` field and Base64 decode it into a byte sequence
    1. if the result is not an sequence of the length 32, then the verification payload is corrupted and the event MUST be discarded
8. create verification material from the event according to the signing [Protocol](#protocol)
9. compare the newly created verification material to that in the envelope
    1. if the values are not equal, the event has been modified in transit and MUST be discarded
10. the event is returned as verified successfully.

> Upon verification of an CloudEvent, implementations MUST return a new event containing only verified data: the Context Attributes (REQUIRED and OPTIONAL) plus the event data. Extension Context Attributes MUST NOT be included in the verified event. This ensures clear separation between verified and unverified data. Users handle either a complete unverified event or a verified event with only verified values—never a mixture of both.

Reviewer 3: No, this isn't about audits or chains... it's about situations where there are environments where messages are sent thru middleware that will sign/verify messages with out the apps on either side knowing about it. Which means if there are multiple layers of that middleware then the design of this needs to support the idea of signing a message that has already been signed. And, of course, this will then also deal with cases where the middleware adds new attributes to the messages that need to be signed

is used for creating and verifying signatures. This extension intentionally
avoids prescribing specific approaches for key management and Public Key
Infrastructure (PKI), as these decisions are highly dependent on organizational
context, existing security infrastructure, and compliance requirements.
Numerous technical solutions exist to make private and public key material
available to producers and consumers.

### Signature

The `VERIFICATION_MATERIAL` of the type
`https://cloudevents.io/verifiability/dsse/v0.1` in the envelope above is
created as follows:

```
SHA256(
  SHA256(UTF8(event.id)) +
  SHA256(UTF8(event.source)) +
  SHA256(UTF8(event.specversion)) +
  SHA256(UTF8(event.type)) +
  SHA256(UTF8(event.datacontenttype)) +
  SHA256(UTF8(event.dataschema)) +
  SHA256(UTF8(event.subject)) +
  SHA256(RFC3339(UTC(event.time))) +
  SHA256(event.data)
)
```

It is the digest of the concatenated digest list of the mandatory Context
Attributes, the OPTIONAL Context Attributes as well as the event data itself.

#### Protocol

This is how to sign a CloudEvent using DSSE:

1. choose a signing key

> 2. create an empty byte sequence

Reviewer 2: I like the fact that each SHA256 is created separately here, then they're all concatenated together. The benefit of that is that when verifying, if I get an event with data before the id, I can still handle it and remember the hash instead of having to retain the complete byte sequence in memory. It may be worth noting this as a design decision.

  4. compute the SHA256 digest of the event's [`type`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#type) Context Attribute in UTF8 and append it to the byte sequence
  5. compute the SHA256 digest of the event's [`datacontenttype`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#datacontenttype) Context Attribute in UTF8 and append it to the byte sequence *(if the attribute is not set, use the digest of the empty byte sequence)*
  6. compute the SHA256 digest of the event's [`dataschema`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#dataschema) Context Attribute in UTF8 and append it to the byte sequence *(if the attribute is not set, use the digest of the empty byte sequence)*
  7. compute the SHA256 digest of the event's [`subject`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#subject) Context Attribute in UTF8 and append it to the byte sequence *(if the attribute is not set, use the digest of the empty byte sequence)*

>   8. compute the SHA256 digest of the event's [`time`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#time) Context Attribute normalized to RFC 3339 Zulu format and append it to the byte sequence (if the attribute is not set, use the digest of the empty byte sequence)

Reviewer 2: This is still an issue, I believe.

The verification material, once Base64 decoded, MUST be a proper DSSE envelope:

```
{
 "payload": "<Base64(VERIFICATION_MATERIAL)>",
 "payloadType": "https://cloudevents.io/verifiability/dsse/v0.1",
 "signatures": [{
  "keyid": "<KEYID>",
  "sig": "<Base64(SIGNATURE)>"
 }]
}
```

The `payloadType` links to the version of this spec that was used to create the
verification material. The spec defines the version of the [DSSE Protocol](https://github.com/secure-systems-lab/dsse/blob/master/protocol.md)
that is to be used.

## Attributes

This extension defines the following attributes:

### dssematerial

- Type: `String`
- Description: The [DSSE JSON Envelope](https://github.com/secure-systems-lab/dsse/blob/master/envelope.md)
  that can be used to verify the authenticity and integrity of the CloudEvent.
- Constraints:
  - REQUIRED
  - If present, MUST be Base64 encoded
  - When decoded, MUST be a valid DSSE JSON Envelope with:
    - `payloadType` of `"https://cloudevents.io/verifiability/dsse/v0.1"`
    - `payload` containing Base64-encoded verification material (32-byte SHA256 digest)
    - `signatures` array with at least one signature object containing `keyid` and `sig` fields

> ### signedextattrs

Reviewer 3: Does the signature include this attribute? I don't think it does, but shouldn't it? Granted if they twiddle the list then I think the verification should fail anyway, but for completeness I wonder if we should include it?

    - Attribute names MUST NOT include `dssematerial` (the verification material attribute)
    - Attribute names MUST NOT include `signedextattrs` itself
    - If any of these rules are violated during verification, the event MUST be discarded
- Example: `"exta,extb,extc"` indicates that extension attributes `exta`, `extb`, and `extc` were signed

## Implementation

[Version 1.0.2 of the DSSE Protocol](https://github.com/secure-systems-lab/dsse/blob/v1.0.2/protocol.md)
is used for creating and verifying signatures. This extension intentionally
avoids prescribing specific approaches for key management and Public Key
Infrastructure (PKI), as these decisions are highly dependent on organizational
context, existing security infrastructure, and compliance requirements.
Numerous technical solutions exist to make private and public key material
available to producers and consumers.

### Signature

The `VERIFICATION_MATERIAL` of the type
`https://cloudevents.io/verifiability/dsse/v0.1` in the DSSE envelope described

> in the Overview section is created as follows:

Reviewer 3: ```suggestion
in the Overview section MUST be created as follows:
```

material delivered with the event. If it is valid, the client creates the
`VERIFICATION_MATERIAL` from the event and compares it to the signed
`VERIFICATION_MATERIAL`. If it is the same, then the event is considered verified.

#### Protocol

Here is how to verify a given CloudEvent:

1. obtain list of acceptable key ids for verification
2. read the event’s `dssematerial` [Extension Context Attribute](https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#extension-context-attributes)
    1. if it is not present or empty, the event was not signed and MUST be discarded
3. decode the `dssematerial` as Base64 per [RFC4648](https://tools.ietf.org/html/rfc4648)
    1. if decoding failed, the verification material is corrupted and the event MUST be discarded
4. parse the result into a [JSON DSSE Envelope](https://github.com/secure-systems-lab/dsse/blob/master/envelope.md)
    1. if parsing fails, the verification material is corrupted and the event MUST be discarded
5. follow the [DSSE verification protocol](https://github.com/secure-systems-lab/dsse/blob/master/protocol.md#protocol)
    1. filter signatures by `keyid` from list of acceptable keys from step 1

>     2. if verification fails, the event MUST be discarded

Author: So security really wanted this. I think a consumer may not know what is valid or not if they are reading from some system.

Maybe we change MUST to SHOULD? Thoughts?

6. read the envelope’s `payloadType` value

>     1. if it is not equal to `https://cloudevents.io/verifiability/dsse/v0.1`, the verification payload is unknown and the event MUST be discarded

Reviewer 3: ...verification fails

7. read the envelope’s `payload` field and Base64 decode it into a byte sequence

>     1. if the result is not an sequence of the length 32, then the verification payload is corrupted and the event MUST be discarded

Reviewer 3: ...verification fails

8. read the event's `signedextattrs` [Extension Context Attribute](https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#extension-context-attributes) (see the [`signedextattrs`](#signedextattrs) attribute definition for constraints)
    1. if it is not present or empty, the list of "signed extension attributes" is considered empty
    2. if it is not empty, treat it as a comma separated string of "signed extension attribute names"

>     3. validate that the `signedextattrs` value adheres to the constraints defined in the [`signedextattrs`](#signedextattrs) attribute definition - if any constraint is violated, the event MUST be discarded

Reviewer 3: merge this into step 2, so:
... treat it as a comma separated string of "signed extension attribute names" and validate it adhere to the constraints defined in ....

9. create verification material from the event according to the signing [Protocol](#protocol)
10. compare the newly created verification material to that in the envelope

>     1. if the values are not equal, the event has been modified in transit and MUST be discarded

Author: Same here

    - Attribute names MUST NOT include [core context attributes](https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#context-attributes) (`id`, `source`, `specversion`, `type`, `datacontenttype`, `dataschema`, `subject`, `time`)
    - Attribute names MUST NOT include `dssematerial` (the verification material attribute)
    - Attribute names MUST NOT include `signedextattrs` itself
    - If any of these rules are violated during verification, the event MUST be discarded
- Example: `"exta,extb,extc"` indicates that extension attributes `exta`, `extb`, and `extc` were signed

## Implementation

[Version 1.0.2 of the DSSE Protocol](https://github.com/secure-systems-lab/dsse/blob/v1.0.2/protocol.md)
is used for creating and verifying signatures. This extension intentionally
avoids prescribing specific approaches for key management and Public Key
Infrastructure (PKI), as these decisions are highly dependent on organizational
context, existing security infrastructure, and compliance requirements.
Numerous technical solutions exist to make private and public key material
available to producers and consumers.

### Signature

The `VERIFICATION_MATERIAL` of the type

> `https://cloudevents.io/verifiability/dsse/v0.1` in the envelope above is

Author: It's the reference to the above example on line 89

    4. compute the SHA256 digest of the event's [`type`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#type) Context Attribute in UTF8 and append it to the byte sequence *(if the attribute is not set, use the digest of the empty byte sequence)*
    5. compute the SHA256 digest of the event's [`datacontenttype`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#datacontenttype) Context Attribute in UTF8 and append it to the byte sequence *(if the attribute is not set, use the digest of the empty byte sequence)*
    6. compute the SHA256 digest of the event's [`dataschema`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#dataschema) Context Attribute in UTF8 and append it to the byte sequence *(if the attribute is not set, use the digest of the empty byte sequence)*
    7. compute the SHA256 digest of the event's [`subject`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#subject) Context Attribute in UTF8 and append it to the byte sequence *(if the attribute is not set, use the digest of the empty byte sequence)*

>     8. compute the SHA256 digest of the event's [`time`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#time) Context Attribute normalized to RFC 3339 Zulu format and append it to the byte sequence (if the attribute is not set, use the digest of the empty byte sequence)

Reviewer 3: > For other attributes, the canonical serialized representation defined by the event format applies.

is it inconsistent, or weird, to have some attributes signed pre-serialization and some signed post-serialization? I guess I'm also (indirectly) asking why timestamps are special? And if there's a good reason for them to be special, how do we know there won't be others as people add extensions?

       Note: Time normalization to Zulu format ensures signature verification remains valid even when intermediaries deserialize and reserialize events with different timezone representations of the same timestamp.
    9. compute the SHA256 digest of the event's [`data`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#event-data) and append it to the byte sequence
    10. for each extension attribute in the list from step 2 (in the given order, from lowest to highest index)
        1. compute the SHA256 digest of the extension attribute's name and append it to the byte sequence

>         2. compute the SHA256 digest of the extension attribute's value and append it to the byte sequence

Author: I disagree that this is inherently a protocol brittleness issue, after some thought. The protocol defines a deterministic canonicalization + hashing procedure, and the signature binds that canonical representation.

If an intermediary or SDK heuristically coerces values (e.g., parses a string into a timestamp) and then re-serializes it differently, that is a modification of signed content. In that case, signature verification should fail, by design, because the payload is no longer bitwise/canonically equivalent to what was signed.

Put differently: signature semantics are not “sign the meaning,” they’re “sign the canonical form.” Intermediaries that wish to forward signed envelopes must treat signed fields as immutable with respect to canonicalization. If they want to change representations (even to a semantically-equivalent timestamp), they must resign or persist the original value somewhere.

process consumers can be sure that events were in fact produced by the claimed
producer (authenticity), and that the events were received exactly as they were
sent, and not modified in transit (integrity) without needing to trust any
intermediaries.

The threats addressed by this extension are those of malicious actors
impersonating CloudEvent producers and of malicious actors modifying messages
in transit.

With interoperability in mind, this design opts for simplicity and robustness
wherever possible.

## Non-goals

This extension only applies to individual events. It does not give consumers
any guarantees about the completeness of the event stream or the order in
which events are delivered. Solutions for these other issues are outside the
scope of this proposal.

Because the CloudEvents specification [requires](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#id)
the combination of event `source` and `id` to be unique per event, signature
replay attacks (where an attacker resubmits a legitimately signed event) are
out of scope for this proposal - the existing source+id uniqueness requirement
provides sufficient protection.

> 

Reviewer 2: Nit: remove redundant blank line.

Further, this extension only aims at *verifiability*. It does not aim to
enable *confidentiality*. Consequently, it does not address the threat of
unauthorized parties being able to read CloudEvents that were not meant for
them (see [Privacy & Security](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#privacy-and-security)
in the CloudEvents spec).

While the design in this extension *can* be used by authorized intermediaries
to modify and re-sign events, it explicitly does not aim to provide a
cryptographic audit trail of event modifications.

> As a general principle, this extension aims to avoid cryptographic agility in

Reviewer 2: I'm not sure what "cryptographic agility" is - is that a term most folks in the relevant audience would understand? (It's entirely possible.)

favor of simplicity.

## Constraints

The following constraints apply to the proposed design:

**Verifiability MUST be OPTIONAL:** This ensures that the additional burden of
producing verification material and performing verification only applies when
verifiability is desired, which is not always the case. Consumers that don't
understand or care about signatures can simply ignore these extension fields
and process events normally.

**The design MUST be backward compatible:** Backward compatibility ensures that
producers can produce verifiable events without any knowledge about whether the
consumers have been configured to and are able to verify events. Consumers
that do not support verification can consume signed events as if they were
unsigned.

**The verification material MUST be contained in the same message as the
event:** The design aims to be simple and robust, and so the verification
material MUST be transported and delivered along with the event that it
describes and not in separate events or even through different channels.

> 

Reviewer 2: Nit: extra blank line.

verify those signatures. The verification material is transported in an
extension attribute alongside the event data.

In a typical flow (using an SDK):
The producer passes a CloudEvent to the SDK, which creates the verification
material and adds it to the CloudEvent. When the consumer's CloudEvents SDK
receives a message with event and verification material, it performs a
verification of the signature against the key and passes on a verified event to
the consumer:

![An illustration showing how the producer's SDK signs an event, and the consumer's SDK verifies it](verifiability1.png)

The verification material is transported in an [Extension Context Attribute](https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#extension-context-attributes)
called `dssematerial` (see [Attributes](#attributes) section below).

The verification material, once Base64 decoded, MUST be a proper DSSE envelope:

```
{
 "payload": "<Base64(VERIFICATION_MATERIAL)>",
 "payloadType": "https://cloudevents.io/verifiability/dsse/v0.1",
 "signatures": [{
  "keyid": "<KEYID>",
  "sig": "<Base64(SIGNATURE)>"
 }]
}
```

The `VERIFICATION_MATERIAL`, once Base64 decoded, MUST be a JSON object:

```
{

>  "core": "<Base64(core_digest)>",

Reviewer 2: Can we capitalize these for consistency with VERIFICATION_MATERIAL, KEY_ID and SIGNATURE? Aside from anything else, that makes it easier to tell that they're not expected to be attribute names.

 "ext":  "<Base64(ext_digest)>",
 "signedextattrs": ["exta", "extb"]
}
```

Where `core_digest` is a 32-byte SHA256 digest of the core CloudEvent fields and
`ext_digest` is a 32-byte SHA256 digest of the signed extension attribute values.
The `ext` and `signedextattrs` fields are OPTIONAL: they MUST both be present
when extension attributes are signed, and MUST both be absent otherwise.

The `payloadType` links to the version of this spec that was used to create the
verification material. The spec defines the version of the [DSSE Protocol](https://github.com/secure-systems-lab/dsse/blob/master/protocol.md)
that is to be used.

## Attributes

This extension defines the following attribute:

### dssematerial

> - Type: `String`

Reviewer 2: It's not clear to me why this is a string *and* specified as base64. If the intention is to transport arbitrary bytes, I think it should be of type `Binary`, which would happen to be base64-encoded in the JSON CloudEvents format, but not necessarily in other formats.

I think it would be better to make it Binary, and explicitly state that logically it's a string, but binary is used to avoid the restrictions on the String type. Alternatively, would it be reasonable to put the JSON on a single line and have it as a string *without* any additional encoding? I suspect the only character disallowed in https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#type-system that causes a problem at the moment is for the line break.

- Description: The [DSSE JSON Envelope](https://github.com/secure-systems-lab/dsse/blob/master/envelope.md)
  that can be used to verify the authenticity and integrity of the CloudEvent.
- Constraints:
  - OPTIONAL
  - If present, MUST be Base64 encoded

>   - When decoded, MUST be a valid DSSE JSON Envelope with:

Reviewer 2: Is it permitted for other fields to be present in the JSON object?

    - `payloadType` of `"https://cloudevents.io/verifiability/dsse/v0.1"`
    - `payload` containing a Base64-encoded JSON object with:
      - `core`: Base64-encoded 32-byte SHA256 digest of the core CloudEvent fields (REQUIRED)
      - `ext`: Base64-encoded 32-byte SHA256 digest of the signed extension attribute values (OPTIONAL)
      - `signedextattrs`: JSON array of extension attribute name strings that were signed (OPTIONAL)
      - `ext` and `signedextattrs` MUST both be present or both be absent

>       - The `signedextattrs` array MUST adhere to the following rules. If any of these rules are violated, the signing function MUST refuse to produce material and the verifier MUST discard the event:

Reviewer 2: What about including the name of an attribute which isn't actually present in the event?

Reviewer 2: From later on, it looks like that's okay - and could be used to assert that the given attribute is absent/empty - but I think it's worth calling out here.

        - Attribute names MUST NOT contain repetitions

>         - Attribute names MUST NOT include any required or optional [Context Attribute](https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#context-attributes) name (`id`, `source`, `specversion`, `type`, `datacontenttype`, `dataschema`, `subject`, `time`)

Reviewer 2: ```suggestion
        - Attribute names MUST NOT include any [core context attribute](https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#context-attributes) name (`id`, `source`, `specversion`, `type`, `datacontenttype`, `dataschema`, `subject`, `time`)
```

  SHA256(UTF8(event.id)) +
  SHA256(UTF8(event.source)) +
  SHA256(UTF8(event.specversion)) +
  SHA256(UTF8(event.type)) +
  SHA256(UTF8(event.datacontenttype)) +
  SHA256(UTF8(event.dataschema)) +
  SHA256(UTF8(event.subject)) +
  SHA256(RFC3339(UTC(event.time))) +
  SHA256(event.data)
)

ext_digest = SHA256(
  SHA256(event.extensionattr1) +
  SHA256(event.extensionattr2)
)

VERIFICATION_MATERIAL = {
  "core": Base64(core_digest),
  "ext":  Base64(ext_digest),          // omitted if no extension attributes are signed
  "signedextattrs": ["extensionattr1", "extensionattr2"]  // omitted if no extension attributes are signed
}
```

`core_digest` is the digest of the concatenated digest list of the mandatory
Context Attributes, the OPTIONAL Context Attributes, and the event data.
`ext_digest` is the digest of the concatenated value digest list of the
signed extension attributes. The extension attribute names are authenticated
by being included in the `signedextattrs` array inside the signed container.
The `ext` and `signedextattrs` fields are only present when extension
attributes are signed.

#### Protocol

This is how to sign a CloudEvent using DSSE:

1. choose a signing key

> 2. choose the list of "signed extension attributes" (the list MUST adhere to the constraints defined in the [`dssematerial`](#dssematerial) attribute definition: no repetitions, no Context Attribute names, and no `dssematerial`)

Reviewer 2: Note either here or elsewhere that extension attributes with a known value but not a known type (which could happen in some SDK implementations if this is an event which has been received and is being propagated) cannot be signed.

    4. compute the SHA256 digest of the event's [`type`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#type) Context Attribute in UTF8 and append it to the byte sequence *(if the attribute is not set, use the digest of the empty byte sequence)*
    5. compute the SHA256 digest of the event's [`datacontenttype`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#datacontenttype) Context Attribute in UTF8 and append it to the byte sequence *(if the attribute is not set, use the digest of the empty byte sequence)*
    6. compute the SHA256 digest of the event's [`dataschema`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#dataschema) Context Attribute in UTF8 and append it to the byte sequence *(if the attribute is not set, use the digest of the empty byte sequence)*
    7. compute the SHA256 digest of the event's [`subject`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#subject) Context Attribute in UTF8 and append it to the byte sequence *(if the attribute is not set, use the digest of the empty byte sequence)*

>     8. compute the SHA256 digest of the event's [`time`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#time) Context Attribute normalized to RFC 3339 Zulu format and append it to the byte sequence (if the attribute is not set, use the digest of the empty byte sequence)

Reviewer 2: Starting a new comment on this as the others are outdated... I think there are a few aspects here:

- Canonicalization to UTC
- Normalization of precision (potentially, anyway)
- The *precise* format used (not just "what's valid in RFC3339" but a canonical "this is the only permissable result for the input value")

There are valid (according to the CloudEvents spec) "timestamp" values that don't have UTC offsets, so don't actually represent instants in time. I'd propose that trying to sign an event with such a timestamp should fail.

Different SDKs may well have different levels of precision for date/time values. I think we should settle on one, truncate to that, and be very specific about it. If trunction to the second would be acceptable, I think that would be simplest: state that there is *no* subsecond value in the signed material.

So for example, an input timestamp of "2026-05-13T10:49:12.123456+01:00" would *have* to be converted to "2026-05-13T09:49:12Z" and that's what gets signed.

We should specify that this is how time-based extension attributes are signed too.

       Note: Time normalization to Zulu format ensures signature verification remains valid even when intermediaries deserialize and reserialize events with different timezone representations of the same timestamp.
    9. compute the SHA256 digest of the event's [`data`](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#event-data) and append it to the byte sequence
4. compute `core_digest` as the SHA256 digest of the byte sequence from step 3
5. if the list from step 2 is not empty:
    1. create an empty byte sequence for the ext digest
    2. for each extension attribute in the list from step 2 (in the given order, from lowest to highest index)

>         1. compute the SHA256 digest of the extension attribute's value and append it to the byte sequence. Non-string attribute values MUST be serialized to their [CloudEvents string representation](https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#type-system) before hashing. *(if the attribute is not present on the event, use the digest of the empty byte sequence)*

Reviewer 2: I think we need more clarity here. Effectively I'd suggest a whole separate section about how each type is hashed, and then use that for both the core and extension attributes.

(I'd also suggest that when hashing a binary value, there's no point in converting it to base64 first - likewise we could use simple encodings for Integer and Boolean values.)

    3. if an `ext` field is present, Base64 decode it into a byte sequence
        1. if the result is not a sequence of length 32, the verification payload is corrupted and the event MUST be discarded
    4. if a `signedextattrs` field is present, read it as a JSON array of strings (the list of "signed extension attribute names")
        1. validate that the `signedextattrs` value adheres to the constraints defined in the [`dssematerial`](#dssematerial) attribute definition - if any constraint is violated, the event MUST be discarded
    5. if `signedextattrs` is present but the `ext` field is absent from the payload, or vice versa, the event MUST be discarded
7. follow the [DSSE verification protocol](https://github.com/secure-systems-lab/dsse/blob/master/protocol.md#protocol)
    1. filter signatures by `keyid` from list of acceptable keys from step 1
    2. if verification fails, the event MUST be discarded
8. recompute `core_digest` from the event according to steps 3-4 of the signing [Protocol](#protocol) and compare it to the `core` byte sequence from step 6.2
    1. if the values are not equal, the event has been modified in transit and MUST be discarded

> 9. if the `ext` field is present (step 6.3) and the consumer wishes to verify extension attributes, recompute `ext_digest` from the event using the `signedextattrs` list from step 6.4 according to step 5 of the signing [Protocol](#protocol) and compare it to the `ext` byte sequence from step 6.3

Reviewer 2: This is where I think we need to be clear that the verifier *must* know the types of the extension attributes in order to verify them. The overall result of verification could be:

- Verified both core and extensions
- Verified core, there were no signed extensions
- Verified code, can't verify extensions due to at least one extension attribute's type not being known
- Verification failed (one of the signatures doesn't match)

*Notes:*

* *In [CloudEvent's type system](https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#type-system) a `Timestamp`'s string encoding is [RFC 3339](https://tools.ietf.org/html/rfc3339). This means that verification of the `time` Context Attribute can only be done with second precision, even though an SDK might allow passing in a timestamp with nanosecond precision.*
* *In [CloudEvent's official Protocol Buffers format](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/formats/cloudevents.proto#L57), the `time` Context Attribute is encoded as a `google.protobuf.Timestamp` and hence does not include time zone information (which RFC 3339 would allow). For interoperability with CloudEvent setups using the Protocol Buffers format, time zone information is ignored in the signing and verification process.*
* *The signature covers the exact byte representation of the event data. Intermediaries that deserialize and reserialize event data (e.g., reformatting JSON whitespace or reordering keys) will invalidate the signature. Implementations that route signed events MUST preserve the original byte sequence of the event data.*

> ## Verification Walkthrough

Reviewer 2: I've stopped reviewing at this point as I suspect changes above will affect the text below.

**The verification material MUST be contained in the same message as the
event:** The design aims to be simple and robust, and so the verification
material MUST be transported and delivered along with the event that it
describes and not in separate events or even through different channels.

## Overview

This extension enables event producers to sign CloudEvents and consumers to
verify those signatures. The verification material is transported in an
extension attribute alongside the event data.

> 

Reviewer 3: extra line

    3. if an `ext` field is present, Base64 decode it into a byte sequence
        1. if the result is not a sequence of length 32, the verification payload is corrupted and the event MUST be discarded
    4. if a `signedextattrs` field is present, read it as a JSON array of strings (the list of "signed extension attribute names")
        1. validate that the `signedextattrs` value adheres to the constraints defined in the [`dssematerial`](#dssematerial) attribute definition - if any constraint is violated, the event MUST be discarded
    5. if `signedextattrs` is present but the `ext` field is absent from the payload, or vice versa, the event MUST be discarded
7. follow the [DSSE verification protocol](https://github.com/secure-systems-lab/dsse/blob/master/protocol.md#protocol)
    1. filter signatures by `keyid` from list of acceptable keys from step 1
    2. if verification fails, the event MUST be discarded
8. recompute `core_digest` from the event according to steps 3-4 of the signing [Protocol](#protocol) and compare it to the `core` byte sequence from step 6.2
    1. if the values are not equal, the event has been modified in transit and MUST be discarded

> 9. if the `ext` field is present (step 6.3) and the consumer wishes to verify extension attributes, recompute `ext_digest` from the event using the `signedextattrs` list from step 6.4 according to step 5 of the signing [Protocol](#protocol) and compare it to the `ext` byte sequence from step 6.3

Reviewer 3: Please wrap entire doc at 80 chars

    1. if the values are not equal, the extension attributes have been modified in transit and MUST be discarded
10. the event is returned as verified successfully.

Upon verification of a CloudEvent, implementations MUST return a new event

> containing only verified data. If only core verification was performed (step 8),

Reviewer 3: So, if there are CE extensions but they're not verified, they're silently dropped?
Is this what we want instead of rejecting the event? I'm wondering if the presence of extra stuff is sign that someone is trying to mess with things and we need to play it safe.

Reviewer 2: I'm concerned about this too. I'd far rather be able to keep everything, but make it easy to determine which extensions were signed and which weren't.
